client/src/command/push.rs

```diff
@@ -41,6 +46,80 @@ pub struct Push {
     force_preamble: bool,
 }
 
+struct PushContext {
+    store: Arc<NixStore>,
+    cache_name: CacheName,
+    server_name: ServerName,
+    pusher: Pusher,
+    no_closure: bool,
+    ignore_upstream_cache_filter: bool,
+}
+
+impl PushContext {
+    async fn push_static(self, paths: &Vec<PathBuf>) -> Result<()> {
+        let roots = paths
+            .clone()
```

Comment:
nit: any reason not to take the owned `Vec` and make callers do the clone for you?